src/lib/mappers/charge-element.js

```diff
@@ -1,65 +1,103 @@
 
 /**
  * Creates a ChargeElement instance given a row of charge element data
  * @param {Object} chargeElementRow - charge element row from the charge processor
  * @return {ChargeElement}
  */
-const dbToModel = row => {
-  const chargeElementRow = camelCaseKeys(row);
-  const element = new ChargeElement();
-  element.fromHash({
-    id: chargeElementRow.chargeElementId,
-    source: chargeElementRow.source,
-    season: chargeElementRow.season,
-    loss: chargeElementRow.loss,
-    abstractionPeriod: abstractionPeriodMapper.dbToModel(chargeElementRow),
-    authorisedAnnualQuantity: chargeElementRow.authorisedAnnualQuantity,
-    billableAnnualQuantity: chargeElementRow.billableAnnualQuantity
-  });
-
-  if (chargeElementRow.description) {
-    element.description = chargeElementRow.description;
-  }
-
-  if (chargeElementRow.purposeUse) {
-    element.purposeUse = purposeUseMapper.dbToModel(chargeElementRow.purposeUse);
-  }
+const dbToModel = row =>
+  helpers.createModel(ChargeElement, camelCaseKeys(row), dbToModelMapper);
 
-  if (chargeElementRow.timeLimitedStartDate && chargeElementRow.timeLimitedEndDate) {
-    element.timeLimitedPeriod = new DateRange(
-      chargeElementRow.timeLimitedStartDate, chargeElementRow.timeLimitedEndDate);
-  }
-
-  return element;
-};
+const pojoToModelMapper = createMapper()
+  .copy(
+    'id',
+    'externalId',
+    'authorisedAnnualQuantity',
+    'billableAnnualQuantity',
+    'source',
+    'season',
+    'loss',
+    'description'
+  )
+  .map('abstractionPeriod').to('abstractionPeriod', abstractionPeriodMapper.pojoToModel)
+  .map('purposePrimary').to('purposePrimary', purposePrimaryMapper.pojoToModel)
+  .map('purposeSecondary').to('purposeSecondary', purposeSecondaryMapper.pojoToModel)
+  .map('purposeUse').to('purposeUse', purposeUseMapper.pojoToModel);
 
 /**
  * Converts a plain object representation of a ChargeElement to a ChargeElement model
  * @param {Object} pojo
  * @return ChargeElement
  */
-const pojoToModel = pojo => {
-  const { abstractionPeriod, purposeUse, ...rest } = pojo;
-  const model = new ChargeElement();
-  model.fromHash(omit(rest, 'eiucSource'));
+const pojoToModel = pojo =>
+  helpers.createModel(ChargeElement, pojo, pojoToModelMapper);
+
+/**
+ * Maps charge element to DB fields
+ * @param {ChargeElement} chargeElement
+ * @param {ChargeVersion} [chargeVersion]
+ * @return {Object}
+ */
+
+const modelToDbMapper = createMapper()
+  .map('id').to('chargeElementId')
+  .copy(
+    'source',
+    'season',
+    'loss',
+    'description',
+    'authorisedAnnualQuantity',
+    'billableAnnualQuantity'
+  )
+  .map('abstractionPeriod.startDay').to('abstractionPeriodStartDay')
+  .map('abstractionPeriod.startMonth').to('abstractionPeriodStartMonth')
+  .map('abstractionPeriod.endDay').to('abstractionPeriodEndDay')
+  .map('abstractionPeriod.endMonth').to('abstractionPeriodEndMonth')
```

Comment:
Thanks, I had a look at this. 

The snag is the object mapper gives you:
```
{
  abstractionPeriod: {
    abstractionPeriodStartDay : 5 ...
```
instead of the desired:
```
{
  abstractionPeriodStartDay : 5...
}
```

I considered adding the ability to specify the root object with the object mapper, e.g. maybe:

```
map('abstractionPeriod').to(null, mappingFunc)
```

But on balance I think I prefer the above, as the abstraction period properties don't have their own DB table and may be mapped differently depending on the table.

I noticed however the abstractionPeriod.modelToDb() wasn't used anywhere else so I removed this.